tephra-core/src/test/java/co/cask/tephra/distributed/ThriftTransactionServerTest.java

```diff
@@ -0,0 +1,236 @@
+    conf.setInt(TxConstants.Service.CFG_DATA_TX_CLIENT_ATTEMPTS, 1);
+    conf.setInt(TxConstants.Service.CFG_DATA_TX_CLIENT_COUNT, NUM_CLIENTS);
+    conf.setLong(TxConstants.Service.CFG_DATA_TX_CLIENT_TIMEOUT, TimeUnit.HOURS.toMillis(1));
+    conf.setInt(TxConstants.Service.CFG_DATA_TX_SERVER_IO_THREADS, 2);
+    conf.setInt(TxConstants.Service.CFG_DATA_TX_SERVER_THREADS, 4);
+
+    injector = Guice.createInjector(
+      new ConfigModule(conf),
+      new ZKModule(),
+      new DiscoveryModules().getDistributedModules(),
+      Modules.override(new TransactionModules().getDistributedModules())
+        .with(new AbstractModule() {
+          @Override
+          protected void configure() {
+            bind(TransactionStateStorage.class).to(SlowTransactionStorage.class).in(Scopes.SINGLETON);
+          }
+        }),
+      new TransactionClientModule()
+    );
+
+    zkClientService = injector.getInstance(ZKClientService.class);
+    zkClientService.startAndWait();
+
+    // start a tx server
+    txService = injector.getInstance(TransactionService.class);
+    storage = injector.getInstance(TransactionStateStorage.class);
+    try {
+      LOG.info("Starting transaction service");
+      txService.startAndWait();
+    } catch (Exception e) {
+      LOG.error("Failed to start service: ", e);
+    }
+  }
+
+  @Before
+  public void reset() throws Exception {
+    getClient().resetState();
+  }
+
+  @AfterClass
+  public static void stop() throws Exception {
+    txService.stopAndWait();
+    storage.stopAndWait();
+    zkClientService.stopAndWait();
+    zkServer.stopAndWait();
+  }
+
+  public TransactionSystemClient getClient() throws Exception {
+    return injector.getInstance(TransactionSystemClient.class);
+  }
+
+  @Test
+  public void testThriftServerStop() throws Exception {
+    int nThreads = NUM_CLIENTS;
+    ExecutorService executorService = Executors.newFixedThreadPool(nThreads);
+    for (int i = 0; i < nThreads; ++i) {
+      executorService.submit(new Runnable() {
+        @Override
+        public void run() {
+          try {
+            TransactionSystemClient client1 = getClient();
+            CLIENTS_DONE_LATCH.countDown();
+            client1.startShort();
+          } catch (Exception e) {
+            // Exception expected
+          }
+        }
+      });
+    }
+
+    // Wait till all clients finish sending reqeust to transaction manager
+    CLIENTS_DONE_LATCH.await();
+    TimeUnit.SECONDS.sleep(1);
+
+    // Expire zookeeper session, which causes Thrift server to stop.
+    expireZkSession(zkClientService);
+    waitForThriftTermination();
+
+    // Stop Zookeeper client so that it does not re-connect to Zookeeper and start Thrift sever again.
+    zkClientService.stopAndWait();
+    STORAGE_WAIT_LATCH.countDown();
+    TimeUnit.SECONDS.sleep(1);
+
+    // Make sure Thrift server stopped.
+    Assert.assertEquals(Service.State.TERMINATED, txService.thriftRPCServerState());
```

Comment:
why not move this assertion into the `waitForThriftTermination` method?
Or are you hoping that in the meantime (while the zkClientService shuts down), that the zkClientService will shift into the TERMINATED state?